src/Nethermind/Nethermind.Serialization.Rlp/Rlp.cs

```diff
@@ -1093,6 +1105,11 @@ public UInt256 DecodeUInt256(int length = -1)
 
                 if (length == -1)
                 {
+                    if (IsAllZeroes(byteSpan))
+                    {
+                        return 0;
```

Comment:
UInt256.Zero

Reply:
The context of this issue was:
basically I save block and force it as main
```
                    env.BlockTree.SuggestBlock(processedBlock, BlockTreeSuggestOptions.ForceSetAsMain);
                    env.BlockTree.UpdateHeadBlock(processedBlock.Hash!);
```
in `UpdateHeadBlock` it fails with `ThrowNonCanonicalUInt256`
given the `decoderContext` description of

```
f901f4a08f64c4436f7213cfdf02cfb9f45d012f1774dfb329b8803de5e7479b11586902a01dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347940000000000000000000000000000000000000000a0a65161379e9ddede34b46cdab99d9eeecbbf1911ed55d995f8b7a3e9008e781f80a04bc1363d91d33e901024d08a5a79b593b0ae86239f84d047fafc9d7b53ac3449b9010000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000801584023f3e208081d480a00000000000000000000000000000000000000000000000000000000000000000880000000000000000a000000000000000000000000000000000000000000000000000000000000000008080
```
at header decoder on
```
            // if we didn't reach the end of the stream, assume we have basefee to decode
            if (decoderContext.Position != headerCheck)
            {
                blockHeader.BaseFeePerGas = decoderContext.DecodeUInt256();
            }
```

And `devoderContext.Position` is `502` while `headerCheck` is `503` and `decoderContext.Length` is `504` 

so I added "all is zero check"

Reply:
Still this is wrong from RLP spec perspective and is a potential consensus issue, can't be accepted

Reply:
Please take a look at `tmp` Dictionary based stores I implemented=)